src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -318,28 +330,43 @@ protected function instantiateObject(array &$data, $class, array &$context, \Ref
 
                         $params = array_merge($params, $data[$paramName]);
                     }
-                } elseif ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
-                    $params[] = $data[$key];
-                    // don't run set for a parameter passed to the constructor
+
+                    continue;
+                }
+
+                if ($allowed && !$ignored && (isset($data[$key]) || array_key_exists($key, $data))) {
+                    $parameterData = $data[$key];
+                    if (null !== $constructorParameter->getClass()) {
+                        $parameterData = $this->serializer->deserialize($parameterData, $constructorParameter->getClass()->getName(), $format, $context);
+                    }
+
+                    // Don't run set for a parameter passed to the constructor
+                    $params[] = $parameterData;
                     unset($data[$key]);
-                } elseif ($constructorParameter->isDefaultValueAvailable()) {
+
+                    continue;
+                }
+
+                if ($constructorParameter->isDefaultValueAvailable()) {
                     $params[] = $constructorParameter->getDefaultValue();
-                } else {
-                    throw new RuntimeException(
-                        sprintf(
-                            'Cannot create an instance of %s from serialized data because its constructor requires parameter "%s" to be present.',
-                            $class,
-                            $constructorParameter->name
-                        )
-                    );
+
+                    continue;
                 }
+
+                throw new RuntimeException(
+                    sprintf(
+                        'Cannot create an instance of %s from serialized data because its constructor requires parameter "%s" to be present.',
+                        $class,
+                        $constructorParameter->name
+                    )
+                );
             }
 
             if ($constructor->isConstructor()) {
                 return $reflectionClass->newInstanceArgs($params);
-            } else {
```

Comment:
Personally the more indent you have the harder it is to read. And if the choice is to use `if/else/elseif` blocks, then everything should be in an else instead of the return done L311.